-1

**Failed Tests**:   Build 
**Size**: This PR adds an extra 28KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-65fced/42929/summary.html
**COMMIT**: ca4a99f3ba9d71e6973f88e120b7bc44de431bc3
**CMSSW**: CMSSW_14_2_X_2024-11-17-2300/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/46722/42929/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Build

I found compilation error when building: 

<pre>>> Compiling edm plugin src/RecoLocalTracker/SiStripClusterizer/plugins/SiStripClusterizer.cc
/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02864/el8_amd64_gcc12/external/gcc/12.3.1-40d504be6370b5a30e3947a6e575ca28/bin/c++ -c -DGNU_GCC -D_GNU_SOURCE -DTBB_USE_GLIBCXX_VERSION=120301 -DTBB_SUPPRESS_DEPRECATED_MESSAGES -DTBB_PREVIEW_RESUMABLE_TASKS=1 -DTBB_PREVIEW_TASK_GROUP_EXTENSIONS=1 -DBOOST_SPIRIT_THREADSAFE -DPHOENIX_THREADSAFE -DBOOST_MATH_DISABLE_STD_FPCLASSIFY -DBOOST_UUID_RANDOM_PROVIDER_FORCE_POSIX -DDD4HEP_USE_GEANT4_UNITS=1 -DCMSSW_GIT_HASH='CMSSW_14_2_X_2024-11-17-2300' -DPROJECT_NAME='CMSSW' -DPROJECT_VERSION='CMSSW_14_2_X_2024-11-17-2300' -Isrc -Ipoison -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02864/el8_amd64_gcc12/cms/cmssw-patch/CMSSW_14_2_X_2024-11-17-2300/src -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02864/el8_amd64_gcc12/cms/coral/CORAL_2_3_21-30a0a4ea666552305c0953736441148a/include/LCG -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02864/el8_amd64_gcc12/external/dd4hep/v01-29-00-d545e40bf32394b8dd54938932048744/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02864/el8_amd64_gcc12/external/alpaka/1.1.0-aba90e6b0efd37975ff68417e953fa78/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02864/el8_amd64_gcc12/external/pcre/8.43-e34796d17981e9b6d174328c69446455/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02864/el8_amd64_gcc12/external/boost/1.80.0-933d355ba5f1bc30f82ea288aaa7b265/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02864/el8_amd64_gcc12/external/bz2lib/1.0.6-d065ccd79984efc6d4660f410e4c81de/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02864/el8_amd64_gcc12/external/clhep/2.4.7.1-17c7283747b6cbc68a3534cb987a890c/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02864/el8_amd64_gcc12/external/cuda/12.4.1-fc5cb0e72dba64b6abbf00089f3a044c/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02864/el8_amd64_gcc12/external/curl/7.79.0-abae79fca02c8e83ab0f428f2a91d96b/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02864/el8_amd64_gcc12/external/gsl/2.6-4e597de26ced42091883404fa44cef2e/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02864/el8_amd64_gcc12/external/hls/2019.08-e6beae7d560007d8bb20c2cf88bfde9a/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02864/el8_amd64_gcc12/external/libuuid/2.34-27ce4c3579b5b1de2808ea9c4cd8ed29/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02864/el8_amd64_gcc12/lcg/root/6.30.09-e1b7daf5388085ced3b34158ddbf7faa/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02864/el8_amd64_gcc12/external/tbb/v2021.9.0-e785b749a0b6cb9c66dc1d78066210e0/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02864/el8_amd64_gcc12/cms/vdt/0.4.3-9de12f50f7b2377a9b0ed85f15b462a9/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02864/el8_amd64_gcc12/external/xerces-c/3.1.3-c7b88eaa36d0408120f3c29826a04bf6/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02864/el8_amd64_gcc12/external/xz/5.2.5-6f3f49b07db84e10c9be594a1176c114/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02864/el8_amd64_gcc12/external/zlib/1.2.11-1a082fc322b0051b504cc023f21df178/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02864/el8_amd64_gcc12/external/eigen/3bb6a48d8c171cf20b5f8e48bfb4e424fbd4f79e-39786ff94f8aa2f543922cad720e1b32/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02864/el8_amd64_gcc12/external/eigen/3bb6a48d8c171cf20b5f8e48bfb4e424fbd4f79e-39786ff94f8aa2f543922cad720e1b32/include/eigen3 -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02864/el8_amd64_gcc12/external/fmt/10.2.1-deba157dd76fed17ac8b42adc60ff53f/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02864/el8_amd64_gcc12/external/md5/1.0.0-5b594b264e04ae51e893b1d69a797ec6/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02864/el8_amd64_gcc12/external/OpenBLAS/0.3.27-6574458f8642b299ac965c654e194a8b/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02864/el8_amd64_gcc12/external/tinyxml2/6.2.0-68d511689c5dbd9a0d296bbb795fa052/include -O3 -pthread -pipe -Werror=main -Werror=pointer-arith -Werror=overlength-strings -Wno-vla -Werror=overflow -std=c++20 -ftree-vectorize -Werror=array-bounds -Werror=format-contains-nul -Werror=type-limits -fvisibility-inlines-hidden -fno-math-errno --param vect-max-version-for-alias-checks=50 -Xassembler --compress-debug-sections -Wno-error=array-bounds -Warray-bounds -fuse-ld=bfd -march=x86-64-v2 -felide-constructors -fmessage-length=0 -Wall -Wno-non-template-friend -Wno-long-long -Wreturn-type -Wextra -Wpessimizing-move -Wclass-memaccess -Wno-cast-function-type -Wno-unused-but-set-parameter -Wno-ignored-qualifiers -Wno-unused-parameter -Wunused -Wparentheses -Werror=return-type -Werror=missing-braces -Werror=unused-value -Werror=unused-label -Werror=address -Werror=format -Werror=sign-compare -Werror=write-strings -Werror=delete-non-virtual-dtor -Werror=strict-aliasing -Werror=narrowing -Werror=unused-but-set-variable -Werror=reorder -Werror=unused-variable -Werror=conversion-null -Werror=return-local-addr -Wnon-virtual-dtor -Werror=switch -fdiagnostics-show-option -Wno-unused-local-typedefs -Wno-attributes -Wno-psabi -DEIGEN_DONT_PARALLELIZE -DEIGEN_MAX_ALIGN_BYTES=64 -Wno-error=unused-variable -DALPAKA_DEFAULT_HOST_MEMORY_ALIGNMENT=128 -DALPAKA_DISABLE_VENDOR_RNG -DBOOST_DISABLE_ASSERTS -flto=auto -fipa-icf -flto-odr-type-merging -fno-fat-lto-objects -Wodr -fPIC -MMD -MF tmp/el8_amd64_gcc12/src/RecoLocalTracker/SiStripClusterizer/plugins/RecoLocalTrackerSiStripClusterizerPlugins/SiStripClusterizer.cc.d src/RecoLocalTracker/SiStripClusterizer/plugins/SiStripClusterizer.cc -o tmp/el8_amd64_gcc12/src/RecoLocalTracker/SiStripClusterizer/plugins/RecoLocalTrackerSiStripClusterizerPlugins/SiStripClusterizer.cc.o
>> Compiling edm plugin src/RecoLocalTracker/SiStripClusterizer/plugins/SiStripClusterizerConditionsESProducer.cc
/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02864/el8_amd64_gcc12/external/gcc/12.3.1-40d504be6370b5a30e3947a6e575ca28/bin/c++ -c -DGNU_GCC -D_GNU_SOURCE -DTBB_USE_GLIBCXX_VERSION=120301 -DTBB_SUPPRESS_DEPRECATED_MESSAGES -DTBB_PREVIEW_RESUMABLE_TASKS=1 -DTBB_PREVIEW_TASK_GROUP_EXTENSIONS=1 -DBOOST_SPIRIT_THREADSAFE -DPHOENIX_THREADSAFE -DBOOST_MATH_DISABLE_STD_FPCLASSIFY -DBOOST_UUID_RANDOM_PROVIDER_FORCE_POSIX -DDD4HEP_USE_GEANT4_UNITS=1 -DCMSSW_GIT_HASH='CMSSW_14_2_X_2024-11-17-2300' -DPROJECT_NAME='CMSSW' -DPROJECT_VERSION='CMSSW_14_2_X_2024-11-17-2300' -Isrc -Ipoison -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02864/el8_amd64_gcc12/cms/cmssw-patch/CMSSW_14_2_X_2024-11-17-2300/src -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02864/el8_amd64_gcc12/cms/coral/CORAL_2_3_21-30a0a4ea666552305c0953736441148a/include/LCG -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02864/el8_amd64_gcc12/external/dd4hep/v01-29-00-d545e40bf32394b8dd54938932048744/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02864/el8_amd64_gcc12/external/alpaka/1.1.0-aba90e6b0efd37975ff68417e953fa78/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02864/el8_amd64_gcc12/external/pcre/8.43-e34796d17981e9b6d174328c69446455/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02864/el8_amd64_gcc12/external/boost/1.80.0-933d355ba5f1bc30f82ea288aaa7b265/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02864/el8_amd64_gcc12/external/bz2lib/1.0.6-d065ccd79984efc6d4660f410e4c81de/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02864/el8_amd64_gcc12/external/clhep/2.4.7.1-17c7283747b6cbc68a3534cb987a890c/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02864/el8_amd64_gcc12/external/cuda/12.4.1-fc5cb0e72dba64b6abbf00089f3a044c/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02864/el8_amd64_gcc12/external/curl/7.79.0-abae79fca02c8e83ab0f428f2a91d96b/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02864/el8_amd64_gcc12/external/gsl/2.6-4e597de26ced42091883404fa44cef2e/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02864/el8_amd64_gcc12/external/hls/2019.08-e6beae7d560007d8bb20c2cf88bfde9a/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02864/el8_amd64_gcc12/external/libuuid/2.34-27ce4c3579b5b1de2808ea9c4cd8ed29/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02864/el8_amd64_gcc12/lcg/root/6.30.09-e1b7daf5388085ced3b34158ddbf7faa/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02864/el8_amd64_gcc12/external/tbb/v2021.9.0-e785b749a0b6cb9c66dc1d78066210e0/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02864/el8_amd64_gcc12/cms/vdt/0.4.3-9de12f50f7b2377a9b0ed85f15b462a9/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02864/el8_amd64_gcc12/external/xerces-c/3.1.3-c7b88eaa36d0408120f3c29826a04bf6/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02864/el8_amd64_gcc12/external/xz/5.2.5-6f3f49b07db84e10c9be594a1176c114/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02864/el8_amd64_gcc12/external/zlib/1.2.11-1a082fc322b0051b504cc023f21df178/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02864/el8_amd64_gcc12/external/eigen/3bb6a48d8c171cf20b5f8e48bfb4e424fbd4f79e-39786ff94f8aa2f543922cad720e1b32/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02864/el8_amd64_gcc12/external/eigen/3bb6a48d8c171cf20b5f8e48bfb4e424fbd4f79e-39786ff94f8aa2f543922cad720e1b32/include/eigen3 -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02864/el8_amd64_gcc12/external/fmt/10.2.1-deba157dd76fed17ac8b42adc60ff53f/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02864/el8_amd64_gcc12/external/md5/1.0.0-5b594b264e04ae51e893b1d69a797ec6/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02864/el8_amd64_gcc12/external/OpenBLAS/0.3.27-6574458f8642b299ac965c654e194a8b/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02864/el8_amd64_gcc12/external/tinyxml2/6.2.0-68d511689c5dbd9a0d296bbb795fa052/include -O3 -pthread -pipe -Werror=main -Werror=pointer-arith -Werror=overlength-strings -Wno-vla -Werror=overflow -std=c++20 -ftree-vectorize -Werror=array-bounds -Werror=format-contains-nul -Werror=type-limits -fvisibility-inlines-hidden -fno-math-errno --param vect-max-version-for-alias-checks=50 -Xassembler --compress-debug-sections -Wno-error=array-bounds -Warray-bounds -fuse-ld=bfd -march=x86-64-v2 -felide-constructors -fmessage-length=0 -Wall -Wno-non-template-friend -Wno-long-long -Wreturn-type -Wextra -Wpessimizing-move -Wclass-memaccess -Wno-cast-function-type -Wno-unused-but-set-parameter -Wno-ignored-qualifiers -Wno-unused-parameter -Wunused -Wparentheses -Werror=return-type -Werror=missing-braces -Werror=unused-value -Werror=unused-label -Werror=address -Werror=format -Werror=sign-compare -Werror=write-strings -Werror=delete-non-virtual-dtor -Werror=strict-aliasing -Werror=narrowing -Werror=unused-but-set-variable -Werror=reorder -Werror=unused-variable -Werror=conversion-null -Werror=return-local-addr -Wnon-virtual-dtor -Werror=switch -fdiagnostics-show-option -Wno-unused-local-typedefs -Wno-attributes -Wno-psabi -DEIGEN_DONT_PARALLELIZE -DEIGEN_MAX_ALIGN_BYTES=64 -Wno-error=unused-variable -DALPAKA_DEFAULT_HOST_MEMORY_ALIGNMENT=128 -DALPAKA_DISABLE_VENDOR_RNG -DBOOST_DISABLE_ASSERTS -flto=auto -fipa-icf -flto-odr-type-merging -fno-fat-lto-objects -Wodr -fPIC -MMD -MF tmp/el8_amd64_gcc12/src/RecoLocalTracker/SiStripClusterizer/plugins/RecoLocalTrackerSiStripClusterizerPlugins/SiStripClusterizerConditionsESProducer.cc.d src/RecoLocalTracker/SiStripClusterizer/plugins/SiStripClusterizerConditionsESProducer.cc -o tmp/el8_amd64_gcc12/src/RecoLocalTracker/SiStripClusterizer/plugins/RecoLocalTrackerSiStripClusterizerPlugins/SiStripClusterizerConditionsESProducer.cc.o
src/RecoLocalTracker/SiStripClusterizer/plugins/ClustersFromRawProducer.cc: In member function 'virtual void {anonymous}::ClusterFiller::fill(edmNew::DetSetVector<SiStripCluster>::TSFastFiller&) const':
src/RecoLocalTracker/SiStripClusterizer/plugins/ClustersFromRawProducer.cc:410:41: error: cannot convert 'edm::DetSet<SiStripDigi>' to 'uint16_t' {aka 'short unsigned int'}
  410 |             rawAlgos.suppressHybridData(unpDigis, suppDigis, ipair * 2);
      |                                         ^~~~~~~~
      |                                         |
      |                                         edm::DetSet<SiStripDigi>
In file included from src/RecoLocalTracker/SiStripClusterizer/plugins/ClustersFromRawProducer.cc:7:
</pre>

--- a/RecoLocalTracker/SiStripClusterizer/plugins/ClustersFromRawProducer.cc
+++ b/RecoLocalTracker/SiStripClusterizer/plugins/ClustersFromRawProducer.cc
@@ -12,6 +12,7 @@
 #include "DataFormats/FEDRawData/interface/FEDRawDataCollection.h"
 #include "EventFilter/SiStripRawToDigi/interface/SiStripFEDBuffer.h"
 #include "DataFormats/SiStripCommon/interface/SiStripConstants.h"
+#include "DataFormats/SiStripDetId/interface/SiStripDetId.h"
 
 #include "CalibFormats/SiStripObjects/interface/SiStripDetCabling.h"
 
@@ -407,7 +408,9 @@ void ClusterFiller::fill(StripClusterizerAlgorithm::output_t::TSFastFiller& reco
               buffer->channel(fedCh), std::back_inserter(unpDigis), ipair * 256, isNonLite, mode, legacy_, lmode, pCode);
           if (fedchannelunpacker::StatusCode::SUCCESS == st_ch) {
             edm::DetSet<SiStripDigi> suppDigis{id};
-            rawAlgos.suppressHybridData(unpDigis, suppDigis, ipair * 2);
+            unsigned int detId = suppDigis.id;
+            uint16_t maxNStrips = SiStripDetId(detId).numberOfAPVs() * 128;
+            rawAlgos.suppressHybridData(maxNStrips, unpDigis, suppDigis, ipair * 2);
             std::copy(std::begin(suppDigis), std::end(suppDigis), perStripAdder);
           }
         }

--- a/RecoLocalTracker/SiStripZeroSuppression/interface/SiStripRawProcessingAlgorithms.h
+++ b/RecoLocalTracker/SiStripZeroSuppression/interface/SiStripRawProcessingAlgorithms.h
@@ -21,7 +21,8 @@ public:
   void initialize(const edm::EventSetup&);
   void initialize(const edm::EventSetup&, const edm::Event&);
 
-  uint16_t suppressHybridData(const edm::DetSet<SiStripDigi>& inDigis,
+  uint16_t suppressHybridData(const uint16_t maxStrip,
+                              const edm::DetSet<SiStripDigi>& inDigis,
                               edm::DetSet<SiStripDigi>& suppressedDigis,
                               uint16_t firstAPV = 0);
 

--- a/RecoLocalTracker/SiStripZeroSuppression/plugins/SiStripZeroSuppression.cc
+++ b/RecoLocalTracker/SiStripZeroSuppression/plugins/SiStripZeroSuppression.cc
@@ -5,10 +5,12 @@
 #include "DataFormats/Common/interface/Handle.h"
 #include "FWCore/ParameterSet/interface/ParameterSet.h"
 #include "DataFormats/Common/interface/DetSetVector.h"
+#include "DataFormats/SiStripDetId/interface/SiStripDetId.h"
 #include "DataFormats/SiStripDigi/interface/SiStripDigi.h"
 #include "DataFormats/SiStripDigi/interface/SiStripRawDigi.h"
 #include "RecoLocalTracker/SiStripZeroSuppression/interface/SiStripRawProcessingFactory.h"
 #include "FWCore/Utilities/interface/transform.h"
+#include "Geometry/TrackerGeometryBuilder/interface/StripGeomDetUnit.h"
 #include <memory>
 
 SiStripZeroSuppression::SiStripZeroSuppression(edm::ParameterSet const& conf)
@@ -151,8 +153,11 @@ inline void SiStripZeroSuppression::processHybrid(const edm::DetSetVector<SiStri
   for (const auto& inDigis : input) {
     edm::DetSet<SiStripDigi> suppressedDigis(inDigis.id);
 
+    unsigned int detId = inDigis.id;
+    uint16_t maxNStrips = SiStripDetId(detId).numberOfAPVs() * 128;
+
     uint16_t nAPVflagged = 0;
-    nAPVflagged = algorithms->suppressHybridData(inDigis, suppressedDigis);
+    nAPVflagged = algorithms->suppressHybridData(maxNStrips, inDigis, suppressedDigis);
 
     storeExtraOutput(inDigis.id, nAPVflagged);
     if (!suppressedDigis.empty())

--- a/RecoLocalTracker/SiStripZeroSuppression/src/SiStripRawProcessingAlgorithms.cc
+++ b/RecoLocalTracker/SiStripZeroSuppression/src/SiStripRawProcessingAlgorithms.cc
@@ -55,44 +55,76 @@ void SiStripRawProcessingAlgorithms::initialize(const edm::EventSetup& es, const
  *
  * @return number of restored APVs
  */
-uint16_t SiStripRawProcessingAlgorithms::suppressHybridData(const edm::DetSet<SiStripDigi>& hybridDigis,
+uint16_t SiStripRawProcessingAlgorithms::suppressHybridData(const uint16_t maxNStrips,
+                                                            const edm::DetSet<SiStripDigi>& hybridDigis,
                                                             edm::DetSet<SiStripDigi>& suppressedDigis,
                                                             uint16_t firstAPV) {
-  uint16_t nAPVFlagged = 0;
-  auto beginAPV = hybridDigis.begin();
-  const auto indigis_end = hybridDigis.end();
-  auto iAPV = firstAPV;
-  while (beginAPV != indigis_end) {
-    const auto endAPV = std::lower_bound(beginAPV, indigis_end, SiStripDigi((iAPV + 1) * 128, 0));
-    const auto nDigisInAPV = std::distance(beginAPV, endAPV);
+  uint16_t nAPVFlagged = 0;  // Count of flagged APVs
+  auto currentDigi = hybridDigis.begin();
+  const auto endDigi = hybridDigis.end();
+  auto currentAPV = firstAPV;
+
+  // Loop through the APVs in the range
+  while (currentDigi != endDigi) {
+    // Determine the range of digis belonging to the current APV
+    const auto nextAPVBoundary = SiStripDigi((currentAPV + 1) * 128, 0);
+
+    // Reject any APV larger than the max possible
+    if (nextAPVBoundary.strip() > maxNStrips) {
+      edm::LogError("SiStripRawProcessingAlgorithms")
+          << "In DetId " << suppressedDigis.id << " encountered APV boundary with strip number "
+          << nextAPVBoundary.strip() << ", which exceeds the maximum allowed value for this module (" << maxNStrips
+          << "). Exiting loop.";
+      break;
+    }
+
+    const auto nextAPVDigi = std::lower_bound(currentDigi, endDigi, nextAPVBoundary);
+    const auto nDigisInAPV = std::distance(currentDigi, nextAPVDigi);
+
+    // Handle based on the number of digis in the current APV
     if (nDigisInAPV > 64) {
+      // Process hybrid data for noisy APV
       digivector_t workDigis(128, -1024);
-      for (auto it = beginAPV; it != endAPV; ++it) {
-        workDigis[it->strip() - 128 * iAPV] = it->adc() * 2 - 1024;
+
+      // Populate `workDigis` with values from `currentDigi`
+      for (auto it = currentDigi; it != nextAPVDigi; ++it) {
+        workDigis[it->strip() - 128 * currentAPV] = it->adc() * 2 - 1024;
       }
+
+      // Perform pedestal subtraction
       digivector_t workDigisPedSubtracted(workDigis);
-      subtractorCMN->subtract(hybridDigis.id, iAPV, workDigis);
+      subtractorCMN->subtract(hybridDigis.id, currentAPV, workDigis);
+
+      // Inspect and restore digis
       const auto apvFlagged = restorer->inspectAndRestore(
-          hybridDigis.id, iAPV, workDigisPedSubtracted, workDigis, subtractorCMN->getAPVsCM());
+          hybridDigis.id, currentAPV, workDigisPedSubtracted, workDigis, subtractorCMN->getAPVsCM());
       nAPVFlagged += apvFlagged;
-      if (getAPVFlags()[iAPV]) {
-        suppressor->suppress(workDigis, iAPV, suppressedDigis);
-      } else {  // bad APV: more than 64 but not flagged
-        for (uint16_t i = 0; i != 128; ++i) {
+
+      // Process based on the APV flag
+      if (getAPVFlags()[currentAPV]) {
+        // Suppress flagged APVs
+        suppressor->suppress(workDigis, currentAPV, suppressedDigis);
+      } else {
+        // Handle bad APV: not flagged but exceeds threshold
+        for (uint16_t i = 0; i < 128; ++i) {
           const auto digi = workDigisPedSubtracted[i];
           if (digi > 0) {
-            suppressedDigis.push_back(SiStripDigi(iAPV * 128 + i, suppressor->truncate(digi)));
+            suppressedDigis.push_back(SiStripDigi(currentAPV * 128 + i, suppressor->truncate(digi)));
           }
         }
       }
-    } else {  // already zero-suppressed, copy and truncate
-      std::transform(beginAPV, endAPV, std::back_inserter(suppressedDigis), [this](const SiStripDigi inDigi) {
+    } else {
+      // Already zero-suppressed: copy and truncate
+      std::transform(currentDigi, nextAPVDigi, std::back_inserter(suppressedDigis), [this](const SiStripDigi& inDigi) {
         return SiStripDigi(inDigi.strip(), suppressor->truncate(inDigi.adc()));
       });
     }
-    beginAPV = endAPV;
-    ++iAPV;
+
+    // Move to the next APV
+    currentDigi = nextAPVDigi;
+    ++currentAPV;
   }
+
   return nAPVFlagged;
 }
 

